File: python/leetcode/array/1051_height_checker.py
"""
1051. Height Checker (Easy)

Students are asked to stand in non-decreasing order of heights for an annual photo.

Return the minimum number of students not standing in the right positions.  (This is the number of students that must move in order for all students to be standing in non-decreasing order of height.)

Example 1:

Input: [1,1,4,2,1,3]
Output: 3
Explanation: 
Students with heights 4, 3 and the last 1 are not standing in the right positions.
"""


class Solution(object):
    def heightChecker(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        h2 = [item for item in heights]
        h2.sort()
        cnt = 0
        for i in range(len(heights)):
            if heights[i] != h2[i]:
                cnt += 1
        return cnt

File: cc/leetcode/1051_height_checker.cc
class Solution {
public:
    int heightChecker(vector<int>& heights) {
        vector<int> bucket(101, 0);
        for (auto item : heights)
            bucket[item]++;
        // accumulate comparison
        int start = -1, end = -1;
        int res = 0;
        for (int i = 0; i <= 100; ++i) {
            if (bucket[i] > 0) {
                start++;
                end = start + bucket[i] - 1;
                for (int j = start; j <= end; j++) {
                    if (heights[j] != i)
                        res++;
                }
                start = end;
            }
        }
        return res;
    }
};
